src/api_server/src/request/vsock.rs

```diff
@@ -4,34 +4,69 @@
 use super::super::VmmAction;
 use crate::parsed_request::{undeprecated_parsed_request, Deprecation, Error, ParsedRequest};
 use crate::request::Body;
+use logger::{IncMetric, METRICS};
 use vmm::vmm_config::vsock::VsockDeviceConfig;
 
 pub(crate) fn parse_put_vsock(body: &Body) -> Result<(ParsedRequest, Deprecation), Error> {
-    Ok(undeprecated_parsed_request(ParsedRequest::new_sync(
-        VmmAction::SetVsockDevice(
-            serde_json::from_slice::<VsockDeviceConfig>(body.raw()).map_err(Error::SerdeJson)?,
-        ),
-    )))
+    METRICS.put_api_requests.vsock_count.inc();
+    let vsock_cfg = serde_json::from_slice::<VsockDeviceConfig>(body.raw()).map_err(|e| {
+        METRICS.put_api_requests.vsock_fails.inc();
+        Error::SerdeJson(e)
+    })?;
+
+    if vsock_cfg.vsock_id.is_some() {
+        // vsock_id field in request is deprecated.
+        METRICS.deprecated_api.deprecated_http_api_calls.inc();
+        let mut depr = Deprecation::default();
+        depr.set_and_append_message("PUT /vsock: vsock_id field is deprecated.");
+        Ok((
+            ParsedRequest::new_sync(VmmAction::SetVsockDevice(vsock_cfg)),
+            depr,
+        ))
+    } else {
+        Ok(undeprecated_parsed_request(ParsedRequest::new_sync(
+            VmmAction::SetVsockDevice(vsock_cfg),
+        )))
+    }
```

Comment:
I don't think `VsockDeviceConfig` is the right place to have parsing logic. There were many efforts to decouple the `vmm` and `api_server` crates and this would be a step back here. I'd like us to keep only vmm functionality in the `vmm` crate. The versioning of any specific HTTP endpoint would break this. What we should have instead is mirror structs for the counterparts in the `vmm` crate, like a `VsockDeviceConfigParser` in `api_server` for the `VsockDeviceConfig` in `vmm`. Unfortunately, this would require a lot of refactoring and is not feasible at this point in time, but we could track this somewhere as technical debt.

Reply:
Makes sense.